// @azure/arm-dataprotection/dist-esm/src/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { BackupVaultResource as BackupVaultResourceMapper, PatchResourceRequestInput as PatchResourceRequestInputMapper, CheckNameAvailabilityRequest as CheckNameAvailabilityRequestMapper, FeatureValidationRequestBase as FeatureValidationRequestBaseMapper, BaseBackupPolicyResource as BaseBackupPolicyResourceMapper, BackupInstanceResource as BackupInstanceResourceMapper, TriggerBackupRequest as TriggerBackupRequestMapper, ValidateForBackupRequest as ValidateForBackupRequestMapper, AzureBackupRehydrationRequest as AzureBackupRehydrationRequestMapper, AzureBackupRestoreRequest as AzureBackupRestoreRequestMapper, SyncBackupInstanceRequest as SyncBackupInstanceRequestMapper, ValidateRestoreRequestObject as ValidateRestoreRequestObjectMapper, AzureBackupFindRestorableTimeRangesRequest as AzureBackupFindRestorableTimeRangesRequestMapper, ResourceGuardResource as ResourceGuardResourceMapper, PatchResourceGuardInput as PatchResourceGuardInputMapper, ResourceGuardProxyBaseResource as ResourceGuardProxyBaseResourceMapper, UnlockDeleteRequest as UnlockDeleteRequestMapper } from "../models/mappers";
export const accept = {
    parameterPath: "accept",
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Accept",
        type: {
            name: "String"
        }
    }
};
export const $host = {
    parameterPath: "$host",
    mapper: {
        serializedName: "$host",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2023-05-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String"
        }
    }
};
export const subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        serializedName: "subscriptionId",
        required: true,
        type: {
            name: "Uuid"
        }
    }
};
export const resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        constraints: {
            MaxLength: 90,
            MinLength: 1
        },
        serializedName: "resourceGroupName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const vaultName = {
    parameterPath: "vaultName",
    mapper: {
        serializedName: "vaultName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String"
        }
    }
};
export const parameters = {
    parameterPath: "parameters",
    mapper: BackupVaultResourceMapper
};
export const parameters1 = {
    parameterPath: "parameters",
    mapper: PatchResourceRequestInputMapper
};
export const parameters2 = {
    parameterPath: "parameters",
    mapper: CheckNameAvailabilityRequestMapper
};
export const location = {
    parameterPath: "location",
    mapper: {
        serializedName: "location",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const nextLink = {
    parameterPath: "nextLink",
    mapper: {
        serializedName: "nextLink",
        required: true,
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export const operationId = {
    parameterPath: "operationId",
    mapper: {
        serializedName: "operationId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters3 = {
    parameterPath: "parameters",
    mapper: FeatureValidationRequestBaseMapper
};
export const backupPolicyName = {
    parameterPath: "backupPolicyName",
    mapper: {
        serializedName: "backupPolicyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters4 = {
    parameterPath: "parameters",
    mapper: BaseBackupPolicyResourceMapper
};
export const backupInstanceName = {
    parameterPath: "backupInstanceName",
    mapper: {
        serializedName: "backupInstanceName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters5 = {
    parameterPath: "parameters",
    mapper: BackupInstanceResourceMapper
};
export const parameters6 = {
    parameterPath: "parameters",
    mapper: TriggerBackupRequestMapper
};
export const parameters7 = {
    parameterPath: "parameters",
    mapper: ValidateForBackupRequestMapper
};
export const parameters8 = {
    parameterPath: "parameters",
    mapper: AzureBackupRehydrationRequestMapper
};
export const parameters9 = {
    parameterPath: "parameters",
    mapper: AzureBackupRestoreRequestMapper
};
export const parameters10 = {
    parameterPath: "parameters",
    mapper: SyncBackupInstanceRequestMapper
};
export const parameters11 = {
    parameterPath: "parameters",
    mapper: ValidateRestoreRequestObjectMapper
};
export const filter = {
    parameterPath: ["options", "filter"],
    mapper: {
        serializedName: "$filter",
        type: {
            name: "String"
        }
    }
};
export const skipToken = {
    parameterPath: ["options", "skipToken"],
    mapper: {
        serializedName: "$skipToken",
        type: {
            name: "String"
        }
    }
};
export const recoveryPointId = {
    parameterPath: "recoveryPointId",
    mapper: {
        serializedName: "recoveryPointId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const jobId = {
    parameterPath: "jobId",
    mapper: {
        serializedName: "jobId",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters12 = {
    parameterPath: "parameters",
    mapper: AzureBackupFindRestorableTimeRangesRequestMapper
};
export const parameters13 = {
    parameterPath: "parameters",
    mapper: ResourceGuardResourceMapper
};
export const resourceGuardsName = {
    parameterPath: "resourceGuardsName",
    mapper: {
        serializedName: "resourceGuardsName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters14 = {
    parameterPath: "parameters",
    mapper: PatchResourceGuardInputMapper
};
export const requestName = {
    parameterPath: "requestName",
    mapper: {
        serializedName: "requestName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const resourceGuardProxyName = {
    parameterPath: "resourceGuardProxyName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[A-Za-z0-9]*$")
        },
        serializedName: "resourceGuardProxyName",
        required: true,
        type: {
            name: "String"
        }
    }
};
export const parameters15 = {
    parameterPath: "parameters",
    mapper: ResourceGuardProxyBaseResourceMapper
};
export const parameters16 = {
    parameterPath: "parameters",
    mapper: UnlockDeleteRequestMapper
};
