// @azure/arm-networkfunction/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ProvisioningState} that the service accepts. */
export var KnownProvisioningState;
(function (KnownProvisioningState) {
    /** Succeeded */
    KnownProvisioningState["Succeeded"] = "Succeeded";
    /** Updating */
    KnownProvisioningState["Updating"] = "Updating";
    /** Deleting */
    KnownProvisioningState["Deleting"] = "Deleting";
    /** Failed */
    KnownProvisioningState["Failed"] = "Failed";
})(KnownProvisioningState || (KnownProvisioningState = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
export var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (KnownCreatedByType = {}));
/** Known values of {@link IngestionType} that the service accepts. */
export var KnownIngestionType;
(function (KnownIngestionType) {
    /** Ipfix */
    KnownIngestionType["Ipfix"] = "IPFIX";
})(KnownIngestionType || (KnownIngestionType = {}));
/** Known values of {@link SourceType} that the service accepts. */
export var KnownSourceType;
(function (KnownSourceType) {
    /** Resource */
    KnownSourceType["Resource"] = "Resource";
})(KnownSourceType || (KnownSourceType = {}));
/** Known values of {@link EmissionType} that the service accepts. */
export var KnownEmissionType;
(function (KnownEmissionType) {
    /** Ipfix */
    KnownEmissionType["Ipfix"] = "IPFIX";
})(KnownEmissionType || (KnownEmissionType = {}));
/** Known values of {@link DestinationType} that the service accepts. */
export var KnownDestinationType;
(function (KnownDestinationType) {
    /** AzureMonitor */
    KnownDestinationType["AzureMonitor"] = "AzureMonitor";
})(KnownDestinationType || (KnownDestinationType = {}));
/** Known values of {@link ApiVersionParameter} that the service accepts. */
export var KnownApiVersionParameter;
(function (KnownApiVersionParameter) {
    /** TwoThousandTwentyTwo0501 */
    KnownApiVersionParameter["TwoThousandTwentyTwo0501"] = "2022-05-01";
    /** TwoThousandTwentyTwo0801 */
    KnownApiVersionParameter["TwoThousandTwentyTwo0801"] = "2022-08-01";
    /** TwoThousandTwentyTwo1101 */
    KnownApiVersionParameter["TwoThousandTwentyTwo1101"] = "2022-11-01";
})(KnownApiVersionParameter || (KnownApiVersionParameter = {}));
