// @azure/arm-dataprotection/dist-esm/src/operations/operationStatusResourceGroupContext.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/** Class containing OperationStatusResourceGroupContext operations. */
export class OperationStatusResourceGroupContextImpl {
    /**
     * Initialize a new instance of the class OperationStatusResourceGroupContext class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets the operation status for an operation over a ResourceGroup's context.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param operationId
     * @param options The options parameters.
     */
    get(resourceGroupName, operationId, options) {
        return this.client.sendOperationRequest({ resourceGroupName, operationId, options }, getOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/operationStatus/{operationId}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.OperationResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.operationId
    ],
    headerParameters: [Parameters.accept],
    serializer
};
